### PR TITLE
[ENH] Add ability to learn skeleton with latent confounders using potentially d-separated sets (PDS)

### PR DESCRIPTION
**Pull request by adam2392**
Fixes #

Changes proposed in this pull request:

- Adds the `LearnSemiMarkovianSkeleton` class for learning skeletons in Semi-Markovian cases

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Review by darthtrevino — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -5,3 +5,25 @@ build-docs:
 	make -C doc/ clean
 	make -C doc/ html-noplot
 	cd doc/ && make view
```
Comment (darthtrevino) [inline Makefile:7]: I'm of the opinion that these scripted tasks in makefiles should be modeled as Poe scripts in pyproject.toml. Makefiles require another system of build-tools to be present, and it splits up our engineering system logic

**Review by darthtrevino — APPROVED** [review]

File: dodiscover/constraint/config.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from enum import Enum, EnumMeta
+
+
+class MetaEnum(EnumMeta):
```
Comment (darthtrevino) [inline dodiscover/constraint/config.py:4]: Can we add a note about the purpose of this metaclass?

**Review by darthtrevino — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (adam2392) [reply Makefile:7]: Sure that makes sense. Do you know how to add the clean commands to poetry Poe? Would it just be the shell commands?

Comment (adam2392) [reply Makefile:7]: We still need the make file for the docs folder tho I think.

File: dodiscover/constraint/config.py (hunk 2 above)
Comment (adam2392) [reply dodiscover/constraint/config.py:4]: Added a note

File: Makefile (hunk 1 above)
Comment (adam2392) [reply Makefile]: I've removed the `Makefile`, but certain shell-based operators, such as the "|" command does not work inside poe.

File: dodiscover/constraint/skeleton.py [hunk 3]
```diff
@@ -263,10 +280,12 @@ def fit(self, context: Context) -> None:
         context : Context
             A context object.
         """
+        self.context = context.copy()
```
Comment (robertness) [inline dodiscover/constraint/skeleton.py:283]: @darthtrevino I assume this doesn't need to be a [deep copy](https://www.geeksforgeeks.org/copy-python-deep-copy-shallow-copy/)?

File: dodiscover/constraint/skeleton.py [hunk 4]
```diff
@@ -496,3 +513,119 @@ def _postprocess_ci_test(
             self.sep_set_[y_var][x_var].append(set(cond_set))
             return True
         return False
+
+
+class LearnSemiMarkovianSkeleton(LearnSkeleton):
+    """Learning a semi-markovian skeleton.
+
+    This proceeds by learning a skeleton by testing edges with candidate
```
Comment (robertness) [inline dodiscover/constraint/skeleton.py:521]: Is there a reference?

What is a "collider oriented DAG"?

File: dodiscover/constraint/skeleton.py [hunk 5]
```diff
@@ -496,3 +513,119 @@ def _postprocess_ci_test(
...
+    alpha : float, optional
+        The significance level for the conditional independence test, by default 0.05.
+    min_cond_set_size : int
+        The minimum size of the conditioning set, by default 0. The number of variables
+        used in the conditioning set.
+    max_cond_set_size : int, optional
+        Maximum size of the conditioning set, by default None. Used to limit
+        the computation spent on the algorithm.
+    max_combinations : int, optional
+        Maximum number of tries with a conditioning set chosen from the set of possible
```
Comment (robertness) [inline dodiscover/constraint/skeleton.py:544]: Can you rephrase the definition of `max_combinations`.  I think I get it, but the language is awkward

File: dodiscover/constraint/skeleton.py [hunk 6]
```diff
@@ -496,3 +513,119 @@ def _postprocess_ci_test(
...
+        If True (default) will sort the existing adjacencies of each variable by its
+        dependencies from strongest to weakest (i.e. largest CI test statistic value to lowest).
+    max_path_length : int, optional
+        The maximum length of any discriminating path, or None if unlimited.
+    ci_estimator_kwargs : dict
+        Keyword arguments for the ``ci_estimator`` function.
+
+    Notes
+    -----
+    The difference
```
Comment (robertness) [inline dodiscover/constraint/skeleton.py:562]: ?

**Review by robertness — DISMISSED** [review]
approved with minor nits

File: dodiscover/constraint/skeleton.py (hunk 3 above)
Comment (adam2392) [reply dodiscover/constraint/skeleton.py:283]: This creates a deep-copy of the necessary components via the `def copy(self)` function inside `Context` class.

File: dodiscover/constraint/skeleton.py (hunk 4 above)
Comment (adam2392) [reply dodiscover/constraint/skeleton.py:521]: I suppose we can use the RFCI paper https://arxiv.org/abs/1104.5617 as a reference. However, I think the idea of PDS sets arose earlier. I added some notes and documentation describing this.

File: dodiscover/constraint/skeleton.py (hunk 6 above)
Comment (adam2392) [reply dodiscover/constraint/skeleton.py:562]: Good catch. I added documentation now for users.

File: dodiscover/constraint/skeleton.py (hunk 5 above)
Comment (adam2392) [reply dodiscover/constraint/skeleton.py]: Done. Lmk wdyt?

File: dodiscover/constraint/skeleton.py (hunk 3 above)
Comment (darthtrevino) [reply dodiscover/constraint/skeleton.py:289]: FWIW if you treat data structures as immutable, and use immutability-safe patterns like structural sharing for mutation, you can avoid deep-copy situations. Are there any good immutability libraries in the Python ecosystem?

**Comment by adam2392** [comment]
@darthtrevino @robertness can you re-approve? :p

File: dodiscover/constraint/_classes.py [hunk 7]
```diff
@@ -162,10 +158,10 @@ def fit(self, context: Context) -> None:
         self.X_ = self.context_.data
 
         # initialize graph object to apply learning
-        sep_set = self._initialize_sep_sets(self.init_graph_)
+        self.separating_sets_ = self._initialize_sep_sets(self.init_graph_)
 
         # learn skeleton graph and the separating sets per variable
-        graph, sep_set = self.learn_skeleton(self.context_, sep_set)
+        graph, self.separating_sets_ = self.learn_skeleton(self.context_, self.separating_sets_)
```
Comment (darthtrevino) [inline dodiscover/constraint/_classes.py:164]: I've been using leading-underscore to indicate private members - is trailing more idiomatic?

**Review by darthtrevino — APPROVED** [review]
LGTM, can you take a look at the context stuff before merging this?

File: dodiscover/constraint/_classes.py (hunk 7 above)
Comment (adam2392) [reply dodiscover/constraint/_classes.py:164]: I use trailing to follow "fitted" attributes pattern that scikit-learn uses. https://scikit-learn.org/dev/developers/develop.html
